cansniffer.c

```diff
@@ -489,45 +512,78 @@ int handle_keyb(int fd){
 	return 1; /* ok */
 };
 
-int handle_bcm(int fd, long currcms){
+int handle_frame(int fd, long currcms){
+
+	bool match = false;
+	bool rx_changed = false;
+	int nbytes, i, pos = 0;
 
-	int nbytes, id, i;
+	unsigned long id;
 
-	struct {
-		struct bcm_msg_head msg_head;
+//	struct {
+//		struct bcm_msg_head msg_head;
 		struct can_frame frame;
-	} bmsg;
+//	} bmsg;
 
-	if ((nbytes = read(fd, &bmsg, sizeof(bmsg))) < 0) {
-		perror("bcm read");
+	if ((nbytes = read(fd, &frame, sizeof(frame))) < 0) {
+		perror("frame read");
 		return 0; /* quit */
 	}
 
-	id = bmsg.msg_head.can_id;
-	ioctl(fd, SIOCGSTAMP, &sniftab[id].currstamp);
+	id = frame.can_id;
 
-	if (bmsg.msg_head.opcode != RX_CHANGED) {
-		printf("received strange BCM opcode %d!\n", bmsg.msg_head.opcode);
-		return 0; /* quit */
+	for(i = 0; i <= idx; i++){
+		if(id == sniftab[i].current.can_id){
+			pos = i;
+			match = true;
+			break;
+		}
 	}
 
-	if (nbytes != sizeof(bmsg)) {
-		printf("received strange BCM data length %d!\n", nbytes);
+	if(!match){
+		pos = idx++;
+		rx_changed = true;
+	}
+	else {
+		if(frame.can_dlc == sniftab[pos].current.can_dlc)
+			for(i=0; i<frame.can_dlc; i++){
+				if(frame.data[i] != sniftab[pos].current.data[i] ){
+					rx_changed = true;
+					break;
+				}
+			}
+		else
+			rx_changed = true;
+	}
+
+//	if (bmsg.msg_head.opcode != RX_CHANGED) {
+//		printf("received strange BCM opcode %d!\n", bmsg.msg_head.opcode);
+//		return 0; /* quit */
+//	}
+
+	if (nbytes != sizeof(frame)) {
+		printf("received strange data length %d!\n", nbytes);
 		return 0; /* quit */
 	}
 
-	sniftab[id].current = bmsg.frame;
-	for (i=0; i < 8; i++)
-		sniftab[id].marker.data[i] |= sniftab[id].current.data[i] ^ sniftab[id].last.data[i];
+	if(rx_changed == true){
+		ioctl(fd, SIOCGSTAMP, &sniftab[pos].currstamp);
 
-	sniftab[id].timeout = (timeout)?(currcms + timeout):0;
+		sniftab[pos].current = frame;
+		for (i=0; i < 8; i++)
```

Comment:
```
for (i = 0; i < 8; i++)

Reply:
The original code has a lot of suppressed whitespace in the existing for loops.